pkg/cloudprovider/providers/gce/gce_disks.go

```diff
@@ -791,55 +877,52 @@ func (gce *GCECloud) ResizeDisk(diskToResize string, oldSize resource.Quantity,
 func (gce *GCECloud) GetAutoLabelsForPD(name string, zone string) (map[string]string, error) {
 	var disk *GCEDisk
-	var err error
+	var diskErr error
+
+	// Get the disk
 	if zone == "" {
-		// For regional PDs this is fine, but for zonal PDs we would like as far
-		// as possible to avoid this case, because GCE doesn't guarantee that
-		// volumes are uniquely named per region, just per zone. However,
-		// creation of GCE PDs was originally done only by name, so we have to
-		// continue to support that.
-		// However, wherever possible the zone should be passed (and it is
-		// passed for most cases that we can control, e.g. dynamic volume
-		// provisioning).
-		disk, err = gce.GetDiskByNameUnknownZone(name)
-		if err != nil {
-			return nil, err
+		// Check for both regional and zonal disks
+
+		// We would like as far as possible to avoid this case, because GCE doesn't
+		// guarantee that volumes are uniquely named per region, just per zone.
+		// However, creation of GCE PDs was originally done only by name, so we have
+		// to continue to support that.
+		// However, wherever possible the zone should be passed (and it is passed
+		// for most cases that we can control, e.g. dynamic volume provisioning).
+		disk, diskErr = gce.GetDiskByNameUnknownZone(name)
+		_, isDiskNotFoundErr := diskErr.(*DiskNotFoundError)
+		if diskErr != nil && !isDiskNotFoundErr {
+			return nil, diskErr
 		}
-	} else {
-		// We could assume the disks exists; we have all the information we need
-		// However it is more consistent to ensure the disk exists,
-		// and in future we may gather addition information (e.g. disk type, IOPS etc)
-		if utilfeature.DefaultFeatureGate.Enabled(features.GCERegionalPersistentDisk) {
-			zoneSet, err := volumeutil.LabelZonesToSet(zone)
-			if err != nil {
-				glog.Warningf("Failed to parse zone field: %q. Will use raw field.", zone)
-			}
 
-			if len(zoneSet) > 1 {
-				// Regional PD
-				disk, err = gce.getRegionalDiskByName(name)
-				if err != nil {
-					return nil, err
-				}
-			} else {
-				// Zonal PD
-				disk, err = gce.getDiskByName(name, zone)
-				if err != nil {
-					return nil, err
-				}
-			}
-		} else {
-			disk, err = gce.getDiskByName(name, zone)
-			if err != nil {
-				return nil, err
-			}
+		regionalDisk, regionalDiskErr := gce.getRegionalDiskByName(name)
+
+		if disk == nil {
+			disk, diskErr = regionalDisk, regionalDiskErr
+		} else if regionalDisk != nil {
+			return nil, fmt.Errorf("Both regional and regular PDs were found with name %q", name)
+		} // else disk != nil && regionalDisk == nil, use the previously found regular disk.
+
+	} else { // We have the zone info to identify the disk
+		zoneSet, zoneErr := volumeutil.LabelZonesToSet(zone)
+		if zoneErr != nil {
+			glog.Warningf("Failed to parse zone field: %q. Will use raw field.", zone)
 		}
+
+		if len(zoneSet) > 1 {
+			disk, diskErr = gce.getRegionalDiskByName(name)
+		} else { // len(zoneSet) == 1
```

Comment:
Not sure this comment is necessarily true `// len(zoneSet) == 1`. Could be empty.

Reply:
`zone == ""` handles that case

Reply:
ah, thats kind of hard to reconcile. Can you just make this an if/elseif/else statement and if `len(zone)==0` throw some error. You don't really have any guarantees at this point if someone else changes code int he future that they aren't modifying `zone` in between that other if statement and here.

This is currently an implied condition based on code somewhere else which would be easy to change and introduce bugs later on.

Reply:
The problem is `volumeutil.LabelZonesToSet(zone)` throws an error if the input is "". I would've preferred to put everything in a switch statement if that were not the case